src/parser/ion_1_0/text/mod.rs

```diff
@@ -0,0 +1,2614 @@
+                    Err::Incomplete(n) => debug!("{}: Err::Incomplete({:?}) at:", context, n),
+                }
+                Err(e)
+            }
+            Ok((i, v)) => {
+                debug!(" {}: <- {:?}", context, &v);
+                Ok((i, v))
+            }
+        }
+    }
+}
+
+/// Follows the following documents:
+/// Ion Text Encoding: http://amzn.github.io/ion-docs/docs/text.html
+/// Ion Specification: http://amzn.github.io/ion-docs/docs/spec.html
+
+/// TODO: Use FnMut combinators to get rid of the Rc<RefCell<_>>
+type Table = Rc<RefCell<CurrentSymbolTable>>;
+
+/// Parses the top level values of the Ion string.
+///
+/// Some values are delimited by values other than whitespace. For example, (1+1) is four
+/// distinct values. In this case the two values are parsed as a pair. If the value has
+/// whitespace as a delimiter, only that value is parsed before continuing.
+///
+/// The last parsed value allows the final value to be parsed without the same delimiting rules.
+///
+/// Encoding: top_level
+pub struct ValueIterator<'a> {
+    pub(crate) remaining: &'a str,
+    pub(crate) current_table: Table,
+    next: Option<ion::Value>,
+}
+
+impl<'a> ValueIterator<'a> {
+    pub(crate) fn new(ion: &'a str) -> Self {
+        Self {
+            remaining: ion,
+            next: None,
+            current_table: Rc::new(RefCell::new(CurrentSymbolTable::SystemV1)),
+        }
+    }
+
+    fn as_shared_symbol_table<'v>(&self, value: &'v ion::Value) -> Option<&'v ion::Struct> {
+        let table = if let ion::Data::Struct(Some(s)) = &value.value {
+            s
+        } else {
+            return None;
+        };
+
+        let tagged_as_shared_symbol_table: bool =
+            value
+                .annotations
+                .as_ref()
+                .map_or(false, |annotations| match annotations.get(0) {
+                    Some(Some(SymbolToken::Known { text }))
+                        if text == "$ion_shared_symbol_table" =>
+                    {
+                        true
+                    }
+                    _ => false,
+                });
+
+        if tagged_as_shared_symbol_table {
+            Some(table)
+        } else {
+            None
+        }
+    }
+
+    fn as_local_symbol_table<'v>(&self, value: &'v ion::Value) -> Option<&'v ion::Struct> {
+        let table = if let ion::Data::Struct(Some(s)) = &value.value {
+            s
+        } else {
+            return None;
+        };
+
+        let tagged_as_local_symbol_table: bool =
+            value
+                .annotations
+                .as_ref()
+                .map_or(false, |annotations| match annotations.get(0) {
+                    Some(Some(SymbolToken::Known { text })) if text == "$ion_symbol_table" => true,
+                    _ => false,
+                });
+
+        if tagged_as_local_symbol_table {
+            Some(table)
+        } else {
+            None
+        }
+    }
```

Comment:
Nope, just needs a home. I've pulled it out.

Reply:
Perhaps CurrentSymbolTable should `impl TryFrom<ion::Struct>` ?

Reply:
I suppose that idea doesn't really make sense, since the table might end up needing the existing table before it can be fully resolved to a CurrentSymbolTable.

Reply:
I've simplified them a little more as well. They are still nearly duplicates but I'm not sure what the behavior should be if a shared symbol table shows up in an Ion stream. I've added some comments around this in a later commit.

And yeah, the spec has this to say about that:

http://amzn.github.io/ion-docs/guides/symbols-guide.html

> A local symbol table is the current symbol table for the subset of values in the stream that occur between the end of the symbol table struct and either the next Ion version marker (IVM) **or the end of the next local symbol table struct** (whichever comes first). This way, a local symbol table struct may contain SymbolTokens from the current symbol table.